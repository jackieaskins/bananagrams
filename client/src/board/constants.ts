import { TILE_SIZE } from '../tile/constants';

const SCREEN_MULTIPLIER = 1.5;
const MIN_TILES = 25;

const getNumTiles = (len: number): number =>
  Math.max(Math.ceil((len * SCREEN_MULTIPLIER) / TILE_SIZE), MIN_TILES);
export const TILES_PER_ROW = getNumTiles(window.screen.width);
export const TILES_PER_COL = getNumTiles(window.screen.height);

export const BOARD_WIDTH = TILES_PER_ROW * TILE_SIZE + 2;
export const BOARD_HEIGHT = TILES_PER_COL * TILE_SIZE + 2;

export const BORDER_STYLE = '1px solid rgba(173, 216, 230, 0.5)';
