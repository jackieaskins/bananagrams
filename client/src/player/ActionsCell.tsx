import { DeleteOutlined } from '@ant-design/icons';
import { Button } from 'antd';

import { useCurrentPlayer } from '../game/stateHooks';
import { kickPlayer } from '../socket';
import { Player } from './types';

type ActionsCellProps = {
  player: Player;
};

const ActionsCell = ({
  player: { userId },
}: ActionsCellProps): JSX.Element | null => {
  const currentPlayer = useCurrentPlayer();

  if (!currentPlayer) {
    return null;
  }

  const { isAdmin, userId: currentPlayerId } = currentPlayer;

  if (isAdmin && userId !== currentPlayerId) {
    return (
      <Button
        icon={<DeleteOutlined />}
        onClick={() => {
          kickPlayer({ userId });
        }}
        type="text"
      />
    );
  }

  return null;
};

export default ActionsCell;
