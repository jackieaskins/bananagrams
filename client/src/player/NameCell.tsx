import { CrownOutlined, KeyOutlined } from '@ant-design/icons';
import { Space } from 'antd';

import { Player } from '../players/types';

type NameCellProps = {
  player: Player;
};

const NameCell = ({
  player: { isAdmin, isTopBanana, username },
}: NameCellProps): JSX.Element => (
  <Space>
    <span>{username}</span>
    {isAdmin && (
      <span>
        <KeyOutlined />
      </span>
    )}
    {isTopBanana && (
      <span>
        <CrownOutlined />
      </span>
    )}
  </Space>
);

export default NameCell;
