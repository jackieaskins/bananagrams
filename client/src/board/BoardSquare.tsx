import { useRecoilCallback } from 'recoil';

import { BoardPosition } from '../boards/types';
import { useCurrentBoardSquare } from '../game/stateHooks';
import { moveTile } from '../socket';
import Tile from '../tile/Tile';
import { TILE_SIZE } from '../tile/constants';
import { selectedTileState } from '../tile/selectedTileState';
import { BORDER_STYLE } from './constants';

const BoardSquare = ({ row, col }: BoardPosition): JSX.Element => {
  const square = useCurrentBoardSquare({ row, col });

  const handleClick = useRecoilCallback(
    ({ set, snapshot }) => async () => {
      const selectedTile = await snapshot.getPromise(selectedTileState);

      if (selectedTile) {
        moveTile({
          tileId: selectedTile.tile.id,
          fromPosition: selectedTile.boardPosition,
          toPosition: { row, col },
        });

        set(selectedTileState, null);
      }
    },
    [row, col]
  );

  const selectTile = useRecoilCallback(
    ({ set, snapshot }) => async ({ tile }) => {
      const selectedTile = await snapshot.getPromise(selectedTileState);

      if (!selectedTile) {
        set(selectedTileState, { tile, boardPosition: { row, col } });
      }
    },
    [col, row]
  );

  return (
    <div
      css={{
        minWidth: `${TILE_SIZE}px`,
        width: `${TILE_SIZE}px`,
        height: `${TILE_SIZE}px`,
        border: BORDER_STYLE,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
      }}
      onClick={handleClick}
    >
      {square ? (
        <Tile
          id={square.tile.id}
          letter={square.tile.letter}
          onClick={selectTile}
          validationStatus={square.validationStatus}
        />
      ) : null}
    </div>
  );
};

export default BoardSquare;
