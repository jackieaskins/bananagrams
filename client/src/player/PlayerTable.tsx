import { Table, Typography } from 'antd';
import { ColumnType } from 'antd/lib/table';

import { useCurrentPlayer, useGamePlayers } from '../game/stateHooks';
import ActionsCell from './ActionsCell';
import NameCell from './NameCell';
import ReadyCell from './ReadyCell';
import { Player } from './types';

const COLUMN_ALIGN = 'center';
const MAX_PLAYERS = 8;

const { Text } = Typography;

const actionsColumn: ColumnType<Player> = {
  align: COLUMN_ALIGN,
  key: 'actions',
  render: function ActionsRender(_: any, player: Player) {
    return <ActionsCell player={player} />;
  },
};

const PlayerTable = (): JSX.Element => {
  const players = useGamePlayers();
  const isAdmin = !!useCurrentPlayer()?.isAdmin;

  return (
    <Table
      bordered
      columns={[
        {
          title: 'Players',
          children: [
            {
              align: COLUMN_ALIGN,
              key: 'ready',
              title: 'Ready?',
              render: function ReadyRender(_, player) {
                return <ReadyCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              key: 'player',
              title: 'Player',
              render: function PlayerRender(_, player) {
                return <NameCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              dataIndex: 'gamesWon',
              title: 'Games won',
            },
            ...(isAdmin && players.length > 1 ? [actionsColumn] : []),
          ],
        },
      ]}
      dataSource={players}
      footer={() => (
        <div css={{ fontSize: '12px', textAlign: 'center' }}>
          <Text type="secondary">
            The game will start when all players are ready
            <br />
            Current player count: {players.length} / {MAX_PLAYERS}
          </Text>
        </div>
      )}
      pagination={false}
      rowKey="userId"
      size="middle"
    />
  );
};

export default PlayerTable;
