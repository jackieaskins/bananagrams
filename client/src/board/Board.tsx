import { useMemo } from 'react';
import Draggable from 'react-draggable';

import BoardSquare from './BoardSquare';
import {
  BOARD_WIDTH,
  BOARD_HEIGHT,
  BORDER_STYLE,
  TILES_PER_COL,
  TILES_PER_ROW,
} from './constants';

const Board = (): JSX.Element => {
  const boardSquares = useMemo(
    () =>
      [...Array(TILES_PER_COL)].map((_, y) => (
        <div key={y} css={{ display: 'flex' }}>
          {[...Array(TILES_PER_ROW)].map((_, x) => (
            <BoardSquare key={x} />
          ))}
        </div>
      )),
    []
  );

  return (
    <Draggable bounds="parent" cancel=".no-drag">
      <div
        css={{
          cursor: 'move',
          display: 'flex',
          flexDirection: 'column',
          width: `${BOARD_WIDTH}px`,
          height: `${BOARD_HEIGHT}px`,
          border: BORDER_STYLE,
        }}
      >
        {boardSquares}
      </div>
    </Draggable>
  );
};

export default Board;
