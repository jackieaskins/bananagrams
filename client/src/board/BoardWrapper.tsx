import { useWindowSize } from '../util/window';
import Board from './Board';
import { BOARD_HEIGHT, BOARD_WIDTH } from './constants';

const BoardView = (): JSX.Element => {
  const { width: windowWidth, height: windowHeight } = useWindowSize();

  return (
    <div
      css={{
        position: 'fixed',
        left:
          BOARD_WIDTH > windowWidth
            ? windowWidth - BOARD_WIDTH - 5
            : (windowWidth - BOARD_WIDTH) / 2,
        right:
          BOARD_WIDTH > windowWidth
            ? windowWidth - BOARD_WIDTH - 5
            : (windowWidth - BOARD_WIDTH) / 2,
        top:
          BOARD_HEIGHT > windowHeight
            ? windowHeight - BOARD_HEIGHT - 5
            : (windowHeight - BOARD_HEIGHT) / 2,
        bottom:
          BOARD_HEIGHT > windowHeight
            ? windowHeight - BOARD_HEIGHT - 5
            : (windowHeight - BOARD_HEIGHT) / 2,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
      }}
    >
      <Board />
    </div>
  );
};

export default BoardView;
