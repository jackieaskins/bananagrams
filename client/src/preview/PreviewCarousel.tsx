import { Button, Select } from 'antd';
import { memo, useCallback, useEffect, useMemo, useState } from 'react';

import { Board } from '../board/types';
import { Hand } from '../hand/types';
import { Player } from '../player/types';
import Preview from './Preview';

type PreviewCarouselProps = {
  boards: Record<string, Board>;
  hands: Record<string, Hand>;
  players: Array<Player>;
};

const PreviewCarousel = ({
  boards,
  hands,
  players,
}: PreviewCarouselProps): JSX.Element => {
  const [selectedUserId, setSelectedUserId] = useState<string | undefined>(
    players[0]?.userId
  );

  useEffect(() => {
    if (!players.some(({ userId }) => userId === selectedUserId)) {
      setSelectedUserId(players[0]?.userId);
    }
  }, [players, selectedUserId]);

  const board = useMemo(
    () => (selectedUserId ? boards[selectedUserId] : null) ?? {},
    [boards, selectedUserId]
  );
  const hand = useMemo(
    () => (selectedUserId ? hands[selectedUserId] : null) ?? [],
    [hands, selectedUserId]
  );

  const options = useMemo(
    () =>
      players.map(({ userId, username }) => ({
        value: userId,
        label: username,
      })),
    [players]
  );

  const updateSelectedPlayer = useCallback(
    (getNextIndex: (currentIndex: number, players: Player[]) => number) => {
      setSelectedUserId((currentUserId) => {
        const currentIndex = players.findIndex(
          ({ userId }) => userId === currentUserId
        );
        const nextIndex = getNextIndex(currentIndex, players);
        return players[nextIndex].userId;
      });
    },
    [players, setSelectedUserId]
  );

  const handleLeftClick = useCallback(() => {
    updateSelectedPlayer((currentIndex, players) =>
      currentIndex <= 0 ? players.length - 1 : currentIndex - 1
    );
  }, [updateSelectedPlayer]);

  const handleRightClick = useCallback(() => {
    updateSelectedPlayer((currentIndex, players) =>
      currentIndex >= players.length - 1 ? 0 : currentIndex + 1
    );
  }, [updateSelectedPlayer]);

  const updateDisabled = useMemo(() => players.length <= 1, [players]);

  return (
    <div css={{ height: '100%', display: 'flex', flexDirection: 'column' }}>
      <div
        css={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'space-between',
        }}
      >
        <Button
          size="small"
          onClick={handleLeftClick}
          type="text"
          disabled={updateDisabled}
        >
          {'<'}
        </Button>

        <Select
          bordered={false}
          options={options}
          size="small"
          value={selectedUserId}
          onChange={setSelectedUserId}
          disabled={updateDisabled}
        />

        <Button
          size="small"
          onClick={handleRightClick}
          type="text"
          disabled={updateDisabled}
        >
          {'>'}
        </Button>
      </div>

      <Preview board={board} hand={hand} />
    </div>
  );
};

export default memo(PreviewCarousel);
