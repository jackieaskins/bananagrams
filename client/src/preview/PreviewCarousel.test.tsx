import { Button } from 'antd';
import { shallow } from 'enzyme';
import { useState } from 'react';

import { boardSquareFixture } from '../fixtures/board';
import { playerFixture } from '../fixtures/player';
import PreviewCarousel from './PreviewCarousel';

const mockUseState = useState as jest.Mock;
const mockSetSelectedUserId = jest.fn().mockName('mockSetSelectedUserId');
jest.mock('react', () => ({
  ...jest.requireActual<any>('react'),
  useEffect: jest.fn((fn) => fn()),
  useState: jest.fn(),
}));

describe('<PreviewCarousel />', () => {
  const renderComponent = (propOverrides = {}) =>
    shallow(
      <PreviewCarousel
        players={[
          playerFixture({ userId: 'p1' }),
          playerFixture({ userId: 'p2' }),
        ]}
        boards={{
          p1: { '0,0': boardSquareFixture() },
          p2: { '1,1': boardSquareFixture() },
        }}
        hands={{
          p1: [{ id: 'B1', letter: 'B' }],
          p2: [{ id: 'B2', letter: 'B' }],
        }}
        {...propOverrides}
      />
    );

  beforeEach(() => {
    mockUseState.mockImplementation((init) => [init, mockSetSelectedUserId]);
  });

  it('renders carousel with first user selected', () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  it('renders if no players, hands, or boards', () => {
    expect(
      renderComponent({ players: [], hands: {}, boards: {} })
    ).toMatchSnapshot();
  });

  it('updates selected user id if current is invalid', () => {
    mockUseState.mockReturnValue([undefined, mockSetSelectedUserId]);

    renderComponent();

    expect(mockSetSelectedUserId).toHaveBeenCalledWith('p1');
  });

  describe('left click', () => {
    const clickLeftButton = (currentUserId?: string) => {
      renderComponent().find(Button).first().simulate('click');

      return mockSetSelectedUserId.mock.calls[0][0](currentUserId);
    };

    it('goes to the left', () => {
      expect(clickLeftButton('p2')).toEqual('p1');
    });

    it('goes to last player if on first', () => {
      expect(clickLeftButton('p1')).toEqual('p2');
    });

    it('goes to last player if no current selection', () => {
      expect(clickLeftButton()).toEqual('p2');
    });
  });

  describe('right click', () => {
    const clickRightButton = (currentUserId?: string) => {
      renderComponent().find(Button).last().simulate('click');

      return mockSetSelectedUserId.mock.calls[0][0](currentUserId);
    };

    it('goes to the right', () => {
      expect(clickRightButton('p1')).toEqual('p2');
    });

    it('goes to first player if on last', () => {
      expect(clickRightButton('p2')).toEqual('p1');
    });

    it('goes to first player if no current selection', () => {
      expect(clickRightButton()).toEqual('p1');
    });
  });
});
