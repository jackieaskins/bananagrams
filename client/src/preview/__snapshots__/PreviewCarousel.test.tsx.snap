// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PreviewCarousel /> renders carousel with first user selected 1`] = `
.emotion-0vj82fd-PreviewCarousel {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-01wj9bod-PreviewCarousel {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-0"
  >
    <Button
      disabled={true}
      onClick={[Function]}
      size="small"
      type="text"
    >
      &lt;
    </Button>
    <ForwardRef(InternalSelect)
      bordered={false}
      disabled={true}
      onChange={[MockFunction mockSetSelectedUserId]}
      options={
        Array [
          Object {
            "label": "username",
            "value": "p1",
          },
          Object {
            "label": "username",
            "value": "p2",
          },
        ]
      }
      size="small"
      value="p1"
    />
    <Button
      disabled={true}
      onClick={[Function]}
      size="small"
      type="text"
    >
      &gt;
    </Button>
  </div>
  <Preview
    board={
      Object {
        "0,0": Object {
          "tile": Object {
            "id": "A1",
            "letter": "A",
          },
          "validationStatus": "VALID",
        },
      }
    }
    hand={
      Array [
        Object {
          "id": "B1",
          "letter": "B",
        },
      ]
    }
  />
</div>
`;

exports[`<PreviewCarousel /> renders if no players, hands, or boards 1`] = `
.emotion-0vj82fd-PreviewCarousel {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-01wj9bod-PreviewCarousel {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-0"
  >
    <Button
      disabled={false}
      onClick={[Function]}
      size="small"
      type="text"
    >
      &lt;
    </Button>
    <ForwardRef(InternalSelect)
      bordered={false}
      disabled={false}
      onChange={
        [MockFunction mockSetSelectedUserId] {
          "calls": Array [
            Array [
              undefined,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      options={Array []}
      size="small"
    />
    <Button
      disabled={false}
      onClick={[Function]}
      size="small"
      type="text"
    >
      &gt;
    </Button>
  </div>
  <Preview
    board={Object {}}
    hand={Array []}
  />
</div>
`;
